clib/cThread.ml

```diff
@@ -107,3 +107,18 @@ let mask_sigalrm f x =
 
 let create f x =
   Thread.create (mask_sigalrm f) x
+
+(*
+  Atomic mutex lock taken from https://gitlab.com/gadmm/memprof-limits/-/blob/master/src/thread_map.ml#L23-34
+  Critical sections :
+   - Mutex.lock does not poll on leaving the blocking section
+     since 4.12.
+   - Never inline, to avoid theoretically-possible reorderings with
+     flambda.
+     (workaround to the lack of masking)
+*)
+let[@inline never] with_lock m ~scope =
+  let () = Mutex.lock m (* BEGIN ATOMIC *) in
+  match (* END ATOMIC *) scope () with
+  | (* BEGIN ATOMIC *) x -> Mutex.unlock m ; (* END ATOMIC *) x
+  | (* BEGIN ATOMIC *) exception e -> Mutex.unlock m ; (* END ATOMIC *) raise e
```

Comment:
No exninfo?

Reply:
You do not need to reattach the exninfo when re-raising pointers do you? The info is still in the store. Anyway this would lead to a circularity since you want this combinator to protect `Exninfo.iraise`.

Reply:
Unfortunately you need to re-attach; I'd like to get rid of exninfo and use something like https://github.com/ocaml/dune/blob/main/otherlibs/stdune-unstable/exn_with_backtrace.mli , but we didn't have a lot of agreement last time we discussed.

Reply:
I don't understand: this re-raises the same exception with the same backtrace. The store has not changed. What goes wrong?

Reply:
Indeed IIRC the problem here is that there is not guarantee that the compiler will actually insert `%reraise` here, so if it inserts `raise` the backtrace would be lost? This has changed in OCaml versions, but AFAIU the only way to guarantee that `reraise` will be use is to actually use `reraise e` , not `raise e`, which can be done by declaring it as an external.